ex5/y_ogawa/main.py

```diff
@@ -0,0 +1,348 @@
+
+    Args:
+        data (np.ndarray): データ
+        weights (np.ndarray): 混合係数
+        means (np.ndarray): 平均値
+        covariances (np.ndarray): 共分散行列
+
+    Returns:
+        responsibilities (np.ndarray): 負担率
+        responsibilities_sum (np.ndarray): 負担率の合計
+    """
+    n_samples = data.shape[0]
+    n_components = weights.shape[0]
+
+    # 負担率を計算する
+    responsibilities = np.zeros((n_samples, n_components))
+    for k in range(n_components):
+        responsibilities[:, k] = weights[k] * multivariate_normal.pdf(
+            data, mean=means[k], cov=covariances[k]
+        )
+    responsibilities_sum = responsibilities.sum(axis=1, keepdims=True)
+    responsibilities /= responsibilities_sum
+
+    return responsibilities, responsibilities_sum
+
+
+def mstep(data: np.ndarray, responsibilities: np.ndarray) -> np.ndarray:
+    """Mステップの実行.
+
+    Args:
+        data (np.ndarray): データ
+        responsibilities (np.ndarray): 負担率
+
+    Returns:
+        weights (np.ndarray): 再計算後の混合係数
+        means (np.ndarray): 再計算後の平均
+        covariances (np.ndarray): 再計算後の共分散行列
+    """
+    n_samples, n_features = data.shape
+    n_components = responsibilities.shape[1]
+
+    N_k = responsibilities.sum(axis=0)
+
+    means = np.zeros((n_components, n_features))
+    covariances = np.zeros((n_components, n_features, n_features))
+    weights = np.zeros(n_components)
+
+    # 平均の再計算
+    for k in range(n_components):
+        means[k, :] = np.sum(responsibilities[:, k, np.newaxis] * data, axis=0) / N_k[k]
+
+    # 共分散行列の再計算
+    for k in range(n_components):
+        # データポイントと平均の差分を計算
+        diff = data - means[k, :]
+
+        # データポイントと平均の差分を重み付けし、行列の積を計算
+        weighted_diff = np.dot((responsibilities[:, k, np.newaxis] * diff).T, diff)
+
+        # 共分散行列を計算し、N_k で割る
+        covariances[k] = weighted_diff / N_k[k]
+
+    # 混合係数の再計算
+    weights = N_k / np.sum(N_k, axis=0)
+
+    return weights, means, covariances
+
+
+def EMstep(
+    data: np.ndarray, weights: np.ndarray, means: np.ndarray, covariances: np.ndarray
+):
+    """EMアルゴリズムの実行.
+
+    Args:
+        data (np.ndarray): データ
+        weights (np.ndarray): 混合係数
+        means (np.ndarray): 平均
+        covariances (np.ndarray): 共分散行列
+
+    Returns:
+        log_likelihood (list): 対数尤度のリスト
+        weights (np.ndarray): 再計算後の混合係数
+        means (np.ndarray): 再計算後の平均
+        covariances (np.ndarray): 再計算後の共分散行列
+    """
+    n_samples = data.shape[0]
+
+    # EMアルゴリズムの試行回数
+    max_try = 100
+
+    # 対数尤度を格納する
+    log_likelihood_list = []
+
+    for j in range(max_try):
+        # Eステップを実行する
+        responsibilities, responsibilities_sum = estep(
+            data, weights, means, covariances
+        )
+
+        # Mステップを実行する
+        weights, means, covariances = mstep(data, responsibilities)
+
+        log_likelihood = 0
+        for i in range(n_samples):
+            log_likelihood += np.log(responsibilities_sum[i])
+
+        log_likelihood_list.append(log_likelihood[0])
```

Comment:
```suggestion
        log_likelihood = np.sum(np.log(responsibilities_sum))
        log_likelihood_list.append(log_likelihood)
```
np.sumを使うことでfor文による加算を1行にまとめることができます。